### DISCUSSION TITLE
<React/JavaScript>: <Add a TypeScript lesson>

### DISCUSSION THREAD

**Discussion by Dillonpw** [TheOdinProject/curriculum#27208] [Ideas]
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [X] Would you like to work on this issue?

### Describe your suggestion

Unless there's something under construction already, it could be worth while to build a stand alone lesson for TypeScript instead of just the mention in the prop types lesson in the React course. We could offer some original content and examples for it along with the links already in the existing lesson. New devs are likely going to use or at least read TypeScript code in the real world and the lesson where its mentioned now doesn't do this intentionally, but somewhat downplays just how broadly adopted typescript is.

### Path

Node / JS

### Lesson Url

https://www.theodinproject.com/lessons/node-path-react-new-type-checking-with-proptypes

### (Optional) Discord Name

dillonpw

**Comment by JoshDevHub** [comment]
I've converted this to a discussion because this is the type of change that'd need a bit more of a deep dive before we can commit to introducing it in the curriculum.

Some things I'm interested in:
1. Would a single lesson really do TypeScript justice? I'm maybe a bit doubtful, but I'm definitely interested in your vision for that.
2. "somewhat downplays just how broadly adopted typescript is" -- I'm not really picking up on this same subtext, but if you are, maybe something can be added to mention its popularity?
3. Is its popularity *innately* a reason it needs to be taught? Personally I'm a bit unconvinced. If we started getting concrete examples where juniors felt their unfamiliarity with it was costing them in job interviews, then I'd think it a lot more urgent. But pretty much every example I hear is actually the opposite: that employers tend to not care too deeply about TS knowledge in a junior candidate. They instead focus on examining other skills/traits and believe a good candidate can just pick up TS (if it's used) on the job.

**Comment by Dillonpw** [comment]
Thank you for hearing me out of the subject!
I'll respond to each of your points individually as best I can.

1. While a single lesson is certainly not enough to do it justice, I debated just making some resource suggestions but ultimately the lesson where TypeScript is mentioned is meant to discuss a related but different subject. While there could be a full course on the language, in the meantime a single lesson could be worthwhile as a way to "dip your toes" into it. I picture something like a breakdown of the most basic types and syntax, followed by a guided example and finishing with an assignment to refactor one of the earlier projects and allow for code submission. 

2. The 2023 Stack Overflow Developer Survey showed TypeScript ranking 5th overall in language popularity in a survey that also took into consideration whether the respondent categorized themselves as professional, learning or other. In another section of the survey described with a graphic as: "...a visualization that shows the distance between the proportion of respondents who want to use a technology (“desired”) and the proportion of users that have used the same technology in the past year and want to continue using it (“admired”),  Typescript ranked 3rd is the desired category and 4th in the admired, showing there is both excitement to use the language and those who have used it intend to continue with it. 

[Link to the Survey]((https://survey.stackoverflow.co/2023/#technology-most-popular-technologies)l)

3. Ultimately you are correct, its unlikely a lack of knowledge in this particular language is costing junior developer's jobs, and it's likely that most job ready developers should be able to pick up TypeScript without much issue. With that said being exposed to it in the standard TOP format to get learners started and if they choose to keep at it and even adding it to their list of skills/CV is only going to add to their likelihood of finding a job.

**Reply by Dillonpw** [reply]
https://survey.stackoverflow.co/2023/#technology-most-popular-technologies

**Comment by silenceinspace** [comment]
I came across this discussion while browsing through the curriculum's issues and would like to comment on it too.

Well, first of all, I am still studying the JS path (at 92%) so my opinion may not be well-formed yet. With that in mind, I remember reading that survey too and some things felt pretty interesting. Speaking of TypeScript, I feel like there was a massive wave of hype around it 1 or 2 years ago, which even a beginner like myself could feel. Most people mentioned that it is a superior version of JavaScript — JavaScript with types and then some. 

Nevertheless, my excitement for TypeScript was not as strong. You know, with most newer tools to appreciate them, it takes a deeper understanding of how much of the new they actually bring to the table. But having React ahead of me, I think I will develop a better opinion.

I genuinely think that adding a proper lesson on TypeScript could be very beneficial, especially with an assignment to go back and refactor some of the students' previous projects. 

@Dillonpw Do you personally get to work with TypeScript? If you have more experience with it, may I also ask you why it is mentioned in the React course and not in the JavaScript one?

**Reply by Dillonpw** [reply]
Sorry for the delayed response: 
I like to do all my own freelance work and personal projects with TypeScript, purely a personal choice. The reason TypeScript does not come up in the JS course is because it would only complicate the learning process, and is in no way required for the projects. Though later it does come up in the React course inside of a lesson on PropTypes which you'll come up to in the near future. React will throw some errors in your editor when certain values' types are not explicitly defined when using vanilla JavaScript. If you opt for TypeScript though there's no need for the prop types definition in your file since everything will already be typed as you go along!

**Comment by nik-rev** [comment]
I think a typescript course would be good. But a single lesson? I don't think that would be enough.

**Comment by mehkij** [comment]
I'm of the mind that a TypeScript course that comes after the vanilla JS course is definitely necessary. Similarly to the first lesson in the React course, you could warn users that they should complete the JS course first before moving on to the TS course. Preferably the TS course would be before the React and Node course, so that students could integrate TS into their React and Node projects. This would probably require some refactoring of the React and Node lessons in general, but I think the benefits are well worth it.

TS is quickly becoming the industry standard if it hasn't already. A number of job listings are requiring it, and it is becoming less of an optional thing rather than a "nice to know". People who aren't looking to learn to get a job can just skip it anyways if they prefer vanilla JS. However I think that if you're looking to start a career, TS is becoming necessary.

**Reply by nik-rev** [reply]
I think typescript wouldn't fit before React. I think React is more important and fundamental than TypeScript. So it should go after Node course in my opinion.

**Reply by mehkij** [reply]
@nikitarevenco I think TS should go before React for the same reason that JS goes before React. JS is the foundation of React, and TS would be the foundation of React by extension of that if you're deciding to use TS for your projects instead of JS. And since TS is just a superset of JS, it should go in between the JS course and React course.

**Reply by nik-rev** [reply]
Right, but TS would make learning React more complicated. You'd have to figure out not only how to make components, and how React works in general, but doing all of that with type checking. This would just make it more confusing to learn React.  On the other hand, if you already knew the important things about React after completing the course, then you could go back and refactor your React components to typescript one-by-one which is actually the recommended way to learn typescript.

**Reply by mehkij** [reply]
@nikitarevenco You could go back and refactor your projects if you learn TS first anyways though. The course could have you learn TS first, learn React using vanilla JS, and have a section towards the end of the React course talking about strategies to refactor your projects using TS. I think it also makes more sense so to do it that way since that way you can tie TS to React, but not the other way around. It makes it so that you can learn TS independently of React, and have TS-specific projects, instead of having a section in the TS course that has you refactor React projects.

**Reply by JoshDevHub** [reply]
> However I think that if you're looking to start a career, TS is becoming necessary.

Why do you think this?

**Reply by mehkij** [reply]
@JoshDevHub Like I said above, more and more jobs are starting to list it as a requirement rather than a nice to have. And from what I've seen across different communities is that TS is kinda just becoming the standard.

**Reply by JoshDevHub** [reply]
What's listed in a job description is always more of a wish list than something the employer literally *must* have. I've yet to find much evidence of junior candidates actively being rejected because they don't know TypeScript. If/when people start saying things like that, I think the team could be much more motivated to include some TS content.

Two more things:
1. I'm much more of a ruby/rails person, so my opinion on whether TS should be supported by the curriculum at all is ... probably not the best. For this, it'd be best to get some opinions from the JavaScript team. I'll tag them @TheOdinProject/javascript 
2. The proposed structure you have above though is something I'd argue very hard against. This is because the React course is a part of the Ruby/Rails curriculum, and TypeScript is absolutely not a must have for a typical Ruby/Rails job. I (and I expect most other Rails maintainers) would be resistant to converting the React course over to using TS, and collectively, the team would probably be resistant to maintaining two separate React courses. I think if there's going to be TS at all, it'd probably have to come between React and Node in the NodeJS path.

**Reply by mehkij** [reply]
@JoshDevHub

> What's listed in a job description is always more of a wish list than something the employer literally must have. I've yet to find much evidence of junior candidates actively being rejected because they don't know TypeScript. If/when people start saying things like that, I think the team could be much more motivated to include some TS content.

I'd agree that a lot of jobs that list TS are definitely looking at it as a nice to have, but I think that just the fact alone that jobs are starting to list TS more often than not is indicative of the direction we're headed. More and more people are adopting TS; [it peaked at 52.7 million downloads on March 3rd of this year](https://npmtrends.com/typescript). I think in general and for the longevity of the course, it wouldn't make sense _not_ to include TS.

> The proposed structure you have above though is something I'd argue very hard against. This is because the React course is a part of the Ruby/Rails curriculum, and TypeScript is absolutely not a must have for a typical Ruby/Rails job. I (and I expect most other Rails maintainers) would be resistant to converting the React course over to using TS, and collectively, the team would probably be resistant to maintaining two separate React courses. I think if there's going to be TS at all, it'd probably have to come between React and Node in the NodeJS path.

You're right, I hadn't considered the Ruby/Rails path. But I don't think anything I suggested would have any real bearing on the Ruby path to begin with. I might've been unclear about what I suggested, as I just now realized Foundations also has you learn vanilla JS, so I'll reiterate here: the full-stack JS path could have you learn TS before React, and after the JavaScript module. As also suggested in a reply to someone else under this thread, there could just be a section of lessons towards the end of the React course that talks about strategies for refactoring projects using TS. I think this would be fine, as the Ruby learners could literally just skip that section since as you mentioned, TS isn't a must have for a typical Ruby/Rails job.

**Reply by mao-sz** [reply]
About the positioning of any possible TS content within the Node pathway,  if TS was ever to be introduced then pedagogically it would only really makes sense to continue using TS from that point onwards. Curriculum-wise, it would feel *very* strange having a whole section on TS, then suddenly not using it anymore in any of the examples (i.e. reverting to "teaching" JS then just assuming learners will add the TS in themselves).

If TS was to come before React, that would mean you either need to do exactly that (very strange pedagogically), or convert the React course to include TS. That then poses a significant maintenance issue as the React course becomes very much separate for each pathway. That's *significantly* more maintenance work at hand. You'd have to convert the appropriate lessons, then any changes made to one version will have to be assessed for change in the other version, assuming people even remember to assess for that change. 

Which means if TS was to ever come into the curriculum, it would have to be after React, either before Node (meaning the entire Node section ideally would be converted to Node/ExpressTS for the above reason about language switch), or after Node.

But I've shared my thoughts below in a separate message about TS in the curriculum outright at this point in time.

**Reply by mehkij** [reply]
@MaoShizhong okay, _this_ I agree with completely. Thanks for the response. I'll read your other comment on the issue and respond there.

**Comment by nik-rev** [comment]
Alternative suggestion:
- Refactor the [Type Checking with Prop Types](https://www.theodinproject.com/lessons/node-path-react-new-type-checking-with-proptypes) lesson to use TypeScript instead of PropTypes. What this means is that:

- There will need to be a small introduction to typescript and an installation guide with Vite

The only thing the lesson will touch on about typescript will be how to set types for react components. TypeScript has many features and none of them have to be used, for now the learner can just use basic TypeScript. That will be all that could be taught in TOP about TypeScript. Just having the learner have it installed and use it for type checking the react components prop types.

**Comment by mao-sz** [comment]
I am currently somewhat strongly of the opinion that TypeScript is not particularly vital to add into the curriculum at this point. Certainly not in its current form, for the reasons @JoshDevHub mentioned about handling the two different pathways.

**If** TS was to be added to the curriculum, I would think it better between React and Node, and only within the Node pathway. On top of that, it can't just be a "touch on a few things". It will either go through features and concepts in a lot more detail, or not at all - TS is only as powerful as you make it. Do note that a big thing to consider as well is that the inclusion of new content, especially in the form of a new section/course, is a *huge* undertaking that involves a lot of time and coordination from the whole team and any external contributors. Thus even if something does end up on the team's radar, it may well not be something that gets touched for some time due to ongoing work on other areas that need to be completed first.

Even then, I do not feel it is something the curriculum *needs* to teach for the following reasons:
- Post-TOP and/or during serious job searching (e.g. if job hunting before finishing the course), the learner would very likely be in a position where they absolutely do not need a course to teach them TS and instead can learn independently and incrementally.
- What jobs use in production does not necessarily mean the content is something the curriculum *has* to teach. There are many things that entry-level job descriptions will include that are not actually strict requirements, where not meeting one or more of them will be a guaranteed disqualification. While this will always differ from company to company, even recently plenty of people have reported getting jobs with tech stacks (which includes entirely languages and their tools) that they have not ever used before.

TS is a superset of JS. The stronger you are with JS, the stronger you'll be with TS. TS is powerful as heck, but only if you make it so. From what people who have interviewed (whether they've gotten job offers or not), and those in the field have said from whom I've spoken to, many agree that someone qualified to be offered an entry level position absolutely can learn TS or any other part of the job's tech stack on the job. Otherwise, many people would not have been able to get their first jobs working in languages and/or frameworks completely different from the ones they learnt in.

My view is strictly on the side of "is it *vital* that the curriculum teaches this at this point in time?", not about how I feel TS is as a whole (I love TS and have had great success learning it independently and incrementally). I do not believe that as of now, the curriculum *needs* to include TS. There are many things that will be useful in a professional environment that TOP intentionally does not go into because it does not need to - they are absolutely things that by the time they become relevant, a person ought to be more than capable of looking into those things themselves. Stuff like additional frameworks, completely different languages, graphQL, cross-platform development etc.

I would definitely like to hear further opinions from those on the team who have more experience in the professional side, particularly those who themselves are involved in the recruitment process and so have a better understanding of the market, employers' perspectives and thus what is and isn't so vital.

**Reply by mehkij** [reply]
> TS is only as powerful as you make it.

At the risk of seeming pedantic here, I'd like to push back on this point, as you made it twice in this comment. Every single technology taught in TOP is only as powerful as you make it. I can learn all the technologies taught in the course and still make not-so-great websites if I don't understand how to fully utilize what I've been taught. Furthermore, you can create a decent frontend project with only HTML, CSS and JavaScript, so why do so many people use React, or other frontend frameworks? It's because those frameworks enhance the developer experience. In many ways, they make things simpler and offer convenience that simple JS doesn't. I am of the opinion that TS is in a similar category, which brings me to your next point:

> Post-TOP and/or during serious job searching (e.g. if job hunting before finishing the course), the learner would very likely be in a position where they absolutely do not need a course to teach them TS and instead can learn independently and incrementally.

This point doesn't make sense to me. Why have a course at all then? Anyone can go out and learn any of the topics TOP teaches completely independently of this course, that isn't the point. People look to a course for a structured curriculum, and that's what TOP provides for them. _If_ TS is important to learn at this point in time (which I think it is but that's obviously what we're trying to determine here, hence the "if"), I think it's something that should be included in the course.

> What jobs use in production does not necessarily mean the content is something the curriculum has to teach. There are many things that entry-level job descriptions will include that are not actually strict requirements, where not meeting one or more of them will be a guaranteed disqualification. While this will always differ from company to company, even recently plenty of people have reported getting jobs with tech stacks (which includes entirely languages and their tools) that they have not ever used before.

Again, refer to my point above. Yes, TOP doesn't _have_ to teach everything used out in the world, but it also doesn't _have_ to teach anything at all. Once again, that isn't really the point. It's about teaching what is necessary and important, especially if the course is centered around kickstarting a career. TOP can't be expected to include every single technology that every single company uses, that of course isn't possible. However, the course _has_ and _should_ include the technologies that are standard or simply prevalent in the industry's production codebases. This is a decent segue into the final point I'll make in response to your comment:

> Do note that a big thing to consider as well is that the inclusion of new content, especially in the form of a new section/course, is a huge undertaking that involves a lot of time and coordination from the whole team and any external contributors. Thus even if something does end up on the team's radar, it may well not be something that gets touched for some time due to ongoing work on other areas that need to be completed first.

> I do not believe that as of now, the curriculum needs to include TS. There are many things that will be useful in a professional environment that TOP intentionally does not go into because it does not need to - they are absolutely things that by the time they become relevant, a person ought to be more than capable of looking into those things themselves.

[TS is _already_ relevant](https://npmtrends.com/typescript). If we're being honest, I think TOP is a little _late_ in adopting TS in the course, actually. Even if we downplay it and say that it's only _becoming_ relevant, this would be fine because as you said, it takes time to add the course anyways.

**Reply by mao-sz** [reply]
Thank you for your response - definitely some sensible points.

Just want to clarify something first - when I said "by the time they become relevant", I probably should've been clearer that I meant "when they become relevant topics to study in the context of the curriculum/pedagogy". I didn't mean when they become relevant in the professional sphere, which yes, TS has been professionally relevant for a long time.

Also to make clear, my opinion about whether TS should be included within the curriculum or not is based on a position where if this was my own course that I had full control over the pedagogy for. Of course, I am not the only person on this team, neither are my opinions going to be the exact same as every other person on the team. Further, if at some point the team does collectively decide to introduce TS material to the curriculum, then you can be most assured that whatever part I can play in that will be to help it be as high quality as possible. My opinions are solely around its introduction outright - hope that will help clarify some of the perspectives I have and share.

---

There are a couple of things that I personally consider when it comes to whether something might require action:
- How *vital* is it that the action be carried out as a whole? By vital, I'm referring to its pedagogical value. I use the word `pedagogy`/`pedagogical` a lot, but it's the best word I can find for it. Something that is valuable in the professional sphere might be but does not necessarily mean it will hold the same pedagogical value in a curriculum aimed at teaching programming fundamentals to get someone to a position of "has the skillset to qualify for an entry level role".
- How much work will this create, both short and long term? Of course, a *huge* amount of work needs to go into making the course material in the first place. We've had a React revamp that took a long time, and are currently starting a Node revamp that's also going to take a long time - and they're revamps of courses we already have material for. Then even after it's done, there will be additional maintenance work.

So combining the two together, you'd have to assess whether the certain material is pedagogically *vital* enough that it warrants the work. If they are, 100% let's work on them. If not, then no.

Points I've thought about regarding TS' pedagogical value:
- If it was to be introduced into the curriculum, where would it go? I've shared my thoughts about this above, about how it really does not make sense to be before React. I do not think it would need to be before Node either, as that will require converting the Node course to TS - the focus of the Node course should be Node/Express (i.e. JS backend then full-stack integration), not TS.
- Given that either side of Node, it'd be pretty much towards/at the end of the curriculum, at a point where people often do start looking for jobs, do we *need* to introduce more content there, especially if they are things that at that point, people very likely have the skills to learn independently?
- You could make that argument about teaching the backend at that point then. Backend concepts are *very* different - they actually *need* being taught and there would be a huge difference between structured learning for the backend from the bottom up, and learning backend independently from the bottom up.
- At this part of the curriculum, I do not see TS falling into this same category. Everything you have done on the frontend as well as what you will be doing on the backend in JS will contribute towards your TS skill already. The TS things are just extensions off those - concepts that do not necessarily need to be taught specifically (they can be, just not necessarily *need* to be - at least not in the same way as many of the backend concepts taught). Even the TS team advise that learning JS *is* learning TS, with the TS parts just being extensions to it.
- The above ties into your question about "Why have a course at all then?". It's context. When is something taught within someone's learning journey and for what reason is it taught at that point or even taught outright? The course provides those things because they make sense pedagogically *and* have that same value for that part of the learning journey. TS would be subject to the same questions relative to where it would be introduced if so.

So that then ultimately boils back down to "is TS important enough *for the curriculum to create and maintain material for* at this point in time?".

Compare it to other material taught at this point, mainly Node and Express. Will a completely independent learning journey for something having just finished the React course follow the same sort of path if they started learning TS independently vs learning the backend independently? I really do not think they compare at all. 

Backend involves many completely new concepts and tools/techniques that seriously benefit from guided learning and being taught directly. And we've also already identified gaps and areas of improvement in the current course that warrant a complete revamp of it, which is currently underway.

Meanwhile, one of the most common ways advised to starting picking up TS for people with this level of general programming and especially JS experience is to just refactor past projects with TS, and incrementally add in more TS features.

---

Hence my opinion that TS is not *vital* at this point to actually teach, given that it would require significant time to coordinate and get the material written and reviewed, as well as introduce additional maintenance work longer term. It does not seem yet like common practice that people are getting rejected specifically because they have JS but not TS experience, nor would TS-specific concepts be as impactful as backend concepts.

This may well change and if so, of course my opinion may well change accordingly.
Similarly, my opinion may differ if there were not already large projects underway which I feel have far more significant pedagogical value for the curriculum.
But as things stand at the moment, I'm not fully convinced it's vital enough to warrant inclusion and the work required for its inclusion/ongoing maitenance.

**Comment by Dillonpw** [comment]
I would like to just chime in here and say thank you to everyone who has voiced an opinion on this topic. I'm glad I was able to open the discussion and hear what both users/students and active contributors/maintainers have to say. It seems we all agree that TypeScript can be an important tool in a professional setting, but we are still trying to come to a conclusion on whether or not it is _vital_ to the curriculum. It appears we all agree that a TypeScript lesson is not enough, and that a course could be too much, but we are unsure how to progress.

If the goal of the curriculum is to provide only the most essential tools to break into web development, then the curriculum already covers what is needed for someone to go from no experience to a full-stack developer and choose how to advance from that point, though as we've surely noticed there is constant change in this field and new technologies are always being developed. If the decision is made to move forward with a TS course at some point in the future, I imagine this discussion would be a good place to look for individuals interested in contributing to the curriculum with TS lessons.

Once again, thank you all for being so vocal and thorough in your opinions and reasoning
